File: ios/FastVideoCdnRuntime.swift
import Foundation

@MainActor
internal final class FastVideoCdnRuntime {
  static let shared = FastVideoCdnRuntime()

  private struct Health {
    var successes = 0
    var failures = 0
    var consecutiveFailures = 0
    var averageTtffMs = 0.0
    var averageResponseMs = 0.0
  }

  private let core = RNFVFastCoreBridge()
  private var health: [String: Health] = [:]

  private init() {}

  func rank(_ uris: [String]) -> [String] {
    uris.enumerated().sorted { lhs, rhs in
      let l = score(uri: lhs.element)
      let r = score(uri: rhs.element)
      return l == r ? lhs.offset < rhs.offset : l > r
    }.map(\.element)
  }

  func recordFirstFrame(uri: String, ttffMs: Double, responseMs: Double? = nil) {
    let key = origin(uri)
    var item = health[key] ?? Health()
    item.successes += 1
    item.consecutiveFailures = 0
    if ttffMs.isFinite, ttffMs >= 0 { item.averageTtffMs = ema(previous: item.averageTtffMs, sample: ttffMs) }
    let response = responseMs ?? ttffMs
    if response.isFinite, response >= 0 { item.averageResponseMs = ema(previous: item.averageResponseMs, sample: response) }
    health[key] = item
  }

  func recordFailure(uri: String, responseMs: Double = 0) {
    let key = origin(uri)
    var item = health[key] ?? Health()
    item.failures += 1
    item.consecutiveFailures += 1
    if responseMs.isFinite, responseMs >= 0 { item.averageResponseMs = ema(previous: item.averageResponseMs, sample: responseMs) }
    health[key] = item
  }

  func diagnostics(uri: String) -> [String: Any] {
    let key = origin(uri)
    let item = health[key] ?? Health()
    return [
      "origin": key,
      "score": score(item),
      "successes": item.successes,
      "failures": item.failures,
      "consecutiveFailures": item.consecutiveFailures,
      "averageTtffMs": item.averageTtffMs,
      "averageResponseMs": item.averageResponseMs
    ]
  }

  func stats() -> [String: Any] {
    [
      "cdnOriginsTracked": health.count,
      "cdnHealth": health.keys.sorted().map { key -> [String: Any] in
        let item = health[key] ?? Health()
        return ["origin": key, "score": score(item), "failures": item.failures, "successes": item.successes]
      }
    ]
  }

  func clear() { health.removeAll(keepingCapacity: true) }

  private func score(uri: String) -> Double { score(health[origin(uri)] ?? Health()) }

  private func score(_ item: Health) -> Double {
    let total = item.successes + item.failures
    let successRate = total == 0 ? 1.0 : Double(item.successes) / Double(total)
    let errorRate = total == 0 ? 0.0 : Double(item.failures) / Double(total)
    let value = core.cdnHealth(
      withSuccessRate: successRate,
      errorRate: errorRate,
      medianTtffMs: item.averageTtffMs,
      medianResponseMs: item.averageResponseMs,
      consecutiveFailures: item.consecutiveFailures
    ) as? [String: Any] ?? [:]
    return (value["score"] as? NSNumber)?.doubleValue ?? 0
  }

  private func ema(previous: Double, sample: Double) -> Double {
    previous <= 0 ? sample : 0.25 * sample + 0.75 * previous
  }

  private func origin(_ value: String) -> String {
    guard let components = URLComponents(string: value), let scheme = components.scheme, let host = components.host else {
      return "local"
    }
    return "\(scheme.lowercased())://\(host.lowercased())\(components.port.map { ":\($0)" } ?? "")"
  }
}

File: android/src/main/java/com/vivekjm/fastvideo/FastVideoCdnRuntime.kt
package com.vivekjm.fastvideo

import android.net.Uri
import org.json.JSONObject

/** Process-local health model for source/fallback origins. Never exports path/query credentials. */
internal object FastVideoCdnRuntime {
  private data class Health(
    var successes: Int = 0,
    var failures: Int = 0,
    var consecutiveFailures: Int = 0,
    var averageTtffMs: Double = 0.0,
    var averageResponseMs: Double = 0.0
  )

  private val health = mutableMapOf<String, Health>()

  @Synchronized
  fun rank(uris: List<String>): List<String> = uris.withIndex()
    .sortedWith(compareByDescending<IndexedValue<String>> { score(it.value) }.thenBy { it.index })
    .map { it.value }

  @Synchronized
  fun recordFirstFrame(uri: String, ttffMs: Double, responseMs: Double = ttffMs) {
    val item = health.getOrPut(origin(uri)) { Health() }
    item.successes += 1
    item.consecutiveFailures = 0
    if (ttffMs.isFinite() && ttffMs >= 0.0) item.averageTtffMs = ema(item.averageTtffMs, ttffMs)
    if (responseMs.isFinite() && responseMs >= 0.0) item.averageResponseMs = ema(item.averageResponseMs, responseMs)
  }

  @Synchronized
  fun recordFailure(uri: String, responseMs: Double = 0.0) {
    val item = health.getOrPut(origin(uri)) { Health() }
    item.failures += 1
    item.consecutiveFailures += 1
    if (responseMs.isFinite() && responseMs >= 0.0) item.averageResponseMs = ema(item.averageResponseMs, responseMs)
  }

  @Synchronized
  fun diagnostics(uri: String): Map<String, Any?> {
    val key = origin(uri)
    val item = health[key] ?: Health()
    return mapOf(
      "origin" to key,
      "score" to scoreFor(item),
      "successes" to item.successes,
      "failures" to item.failures,
      "consecutiveFailures" to item.consecutiveFailures,
      "averageTtffMs" to item.averageTtffMs,
      "averageResponseMs" to item.averageResponseMs
    )
  }

  @Synchronized
  fun stats(): Map<String, Any> = mapOf(
    "cdnOriginsTracked" to health.size,
    "cdnHealth" to health.entries.sortedBy { it.key }.map { (key, value) ->
      mapOf("origin" to key, "score" to scoreFor(value), "failures" to value.failures, "successes" to value.successes)
    }
  )

  @Synchronized
  fun clear() = health.clear()

  private fun score(uri: String): Double = scoreFor(health[origin(uri)] ?: Health())

  private fun scoreFor(item: Health): Double {
    val total = item.successes + item.failures
    val successRate = if (total == 0) 1.0 else item.successes.toDouble() / total
    val errorRate = if (total == 0) 0.0 else item.failures.toDouble() / total
    return runCatching {
      JSONObject(FastCoreNative.cdnHealth(successRate, errorRate, item.averageTtffMs, item.averageResponseMs, item.consecutiveFailures))
        .optDouble("score", 0.0)
    }.getOrDefault(0.0)
  }

  private fun ema(previous: Double, sample: Double): Double = if (previous <= 0.0) sample else 0.25 * sample + 0.75 * previous

  private fun origin(value: String): String = runCatching {
    val uri = Uri.parse(value)
    if (uri.scheme.isNullOrBlank() || uri.host.isNullOrBlank()) "local" else buildString {
      append(uri.scheme!!.lowercase())
      append("://")
      append(uri.host!!.lowercase())
      if (uri.port > 0) append(":${uri.port}")
    }
  }.getOrDefault("unknown")
}
